

<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">

<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Analysis types &mdash; Qualimap v1.0 documentation</title>
    <link rel="stylesheet" href="_static/agogo.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    '',
        VERSION:     '1.0',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <link rel="top" title="Qualimap v1.0 documentation" href="index.html" />
    <link rel="next" title="Command line interface" href="command_line.html" />
    <link rel="prev" title="Workflow" href="workflow.html" /> 
  </head>
  <body>
    <div class="header-wrapper">
      <div class="header">
          <p class="logo"><a href="index.html">
            <img class="logo" src="_static/qualimap_logo_small.png" alt="Logo"/>
          </a></p>
        <h1><a href="index.html">Qualimap v1.0 documentation</a></h1>
        <div class="rel">
          <a href="workflow.html" title="Workflow"
             accesskey="P">previous</a> |
          <a href="command_line.html" title="Command line interface"
             accesskey="N">next</a>
        </div>
       </div>
    </div>

    <div class="content-wrapper">
      <div class="content">
        <div class="document">
            
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body">
            
  <div class="section" id="analysis-types">
<span id="id1"></span><h1>Analysis types<a class="headerlink" href="#analysis-types" title="Permalink to this headline">¶</a></h1>
<div class="section" id="genomic">
<span id="id2"></span><h2>Genomic<a class="headerlink" href="#genomic" title="Permalink to this headline">¶</a></h2>
<p>Genomic analysis reports information for the evaluation of the quality of the alignment data provided. This information eases decision-making for further analysis of the data. In short, the basic statistics of the alignment (number of reads, coverage, GC content, etc.) are summarized and a number of useful graphs are produced. In addition, the user can provide an annotation file so the results are provided for the reads mapping inside (and optionally outside) of the corresponding genomic regions.</p>
<div class="section" id="input">
<h3>Input<a class="headerlink" href="#input" title="Permalink to this headline">¶</a></h3>
<dl class="docutils">
<dt><em class="guilabel">BAM file</em></dt>
<dd>Path to the sequence alignment file in BAM format. Note, that the BAM file has to be sorted by coordinate. Sorting can be performed with <a class="reference external" href="http://samtools.sourceforge.net/">satmtools</a>.</dd>
<dt><em class="guilabel">Draw chromosome limits</em></dt>
<dd>If selected, vertical dotted lines will be placed at the beginning of each chromosome according to the information found in the header of the BAM file.</dd>
<dt><em class="guilabel">Analyze regions</em></dt>
<dd>Activating this option results in analysis of the alignment data only in specified region.</dd>
<dt><em class="guilabel">GFF file</em></dt>
<dd>The path to the annotation file that specifies regions of interest.</dd>
<dt><em class="guilabel">Analyze outside regions</em></dt>
<dd>This option allows to provide the information about the alignment data outside of the regions of interest. Useful for comparison purposes.</dd>
<dt><em class="guilabel">Compare GC content distribution with...</em></dt>
<dd>Option allows to compare calculated distribution with selected precalculated genome distribution. Currently two genome distributions are available: human (hg19) and mouse (mm9)</dd>
</dl>
<div class="section" id="advanced-parameters">
<h4>Advanced parameters<a class="headerlink" href="#advanced-parameters" title="Permalink to this headline">¶</a></h4>
<dl class="docutils">
<dt><em class="guilabel">Number of windows</em></dt>
<dd>Number of windows used to split the selected genome reference. Default is 400.</dd>
<dt><em class="guilabel">Number of threads</em></dt>
<dd>The genomic analysis computation can be performed in parallel on a multicore system using the given number of threads. The default number of threads equals number of available processors.</dd>
<dt><em class="guilabel">Reads per chunk</em></dt>
<dd>To speed up the computation reads are analyzed in chunks. Each chunk is analyzed by single thread. This option controls the number of reads in a chunk. Smaller number may result in lower performance, but also the memory consumption will be reduced.</dd>
</dl>
</div>
</div>
<div class="section" id="output">
<h3>Output<a class="headerlink" href="#output" title="Permalink to this headline">¶</a></h3>
<dl class="docutils">
<dt><em class="guilabel">Summary</em></dt>
<dd>Basic information and statistics for the alignment sequencing input. For example, number of reads, number of mapped reads, mean coverage, chromosome-based statistics, etc.</dd>
<dt><em class="guilabel">Input data &amp; parameters</em></dt>
<dd>Section provides the information about selected input parameters.</dd>
<dt><em class="guilabel">Coverage arcross reference</em></dt>
<dd>This plot consists of two figures. The upper figure provides the coverage (red color) and coverage deviation across the reference sequence. The lower figure show gc-content across reference (black color) and it’s deviation.</dd>
<dt><em class="guilabel">Coverage</em></dt>
<dd>Histogram Frequency histogram of the coverage.</dd>
<dt><em class="guilabel">Coverage histogram (0-50X)</em></dt>
<dd>There is often big picks of coverage across the reference and the scale of the Coverage Histogram graph scale may not be adequate. In order to solve this, in this graph genome locations with a coverage greater than 50X are groped into the last bin.</dd>
<dt><em class="guilabel">Coverage quota</em></dt>
<dd>Provides an easy way of viewing how much reference has been sequenced with a coverage higher than a selected level.</dd>
<dt><em class="guilabel">Mapped Reads Nucleotide Content</em></dt>
<dd>This plot demonstrates the nucleotide content per read position.</dd>
<dt><em class="guilabel">Mapped Reads GC Content Distribution</em></dt>
<dd>This graph shows the distribution of GC content per read. When compared with genome distribution this plot allows to check if there is a shift in the GC content.</dd>
<dt><em class="guilabel">Duplication Rate Histogram</em></dt>
<dd>The histogram shows how many reads start at unique position. This plot is helpful to see if the fragment distribution across genome.</dd>
<dt><em class="guilabel">Mapping quality across reference</em></dt>
<dd>Plot provides mapping quality across reference as indicate by the parameters.</dd>
<dt><em class="guilabel">Mapping quality histogram</em></dt>
<dd>Histogram of the mapping quality frequency</dd>
</dl>
</div>
</div>
<div class="section" id="rna-seq">
<span id="id3"></span><h2>RNA-seq<a class="headerlink" href="#rna-seq" title="Permalink to this headline">¶</a></h2>
<p>In RNA-seq experiments, the reads are mapped to a reference genome. If the total amount of sequencing reads is enough, the number of those reads mapping to a certain biological feature of interest (gene,
transcript, exon, ...) is an estimation of the abundance of that feature in the sample and can be used
as the quantification of its expression level.</p>
<p>These count data are usually utilized to assess differential expression between two or more experimental conditions. But before computing differential expression, users should be aware of some potential limitations of the RNA-seq data they are working with, as for example: has the saturation been reached or could more features be detected by increasing sequencing  depth? Which kind of features are being detected in the experiment? how good is the quantification of expression in the sample? All of these questions are easily answered by looking at the plots generated by Qualimap.</p>
<p>To study the quality of a sample from the count data in a RNA-seq experiment, please use the
RNA-seq option from the Analysis menu.</p>
<p>For this option to work, the R language must be installed in the user’s computer along with the R library “optparse” (both are freely available from <a class="reference external" href="http://cran.r-project.org/">http://cran.r-project.org/</a> ).</p>
<div class="section" id="id4">
<h3>Input<a class="headerlink" href="#id4" title="Permalink to this headline">¶</a></h3>
<dl class="docutils">
<dt><em class="guilabel">First sample (counts)</em></dt>
<dd>File containing the count data from sample. Count data must be provided in a tab-delimited txt file, with the features names or IDs in the first column and counts in the second column. This file must not contain any header or column names. One can calculate the counts from a GFF file and a BAM file using option <em class="menuselection">Tools ‣ Compute Counts</em></dd>
<dt><em class="guilabel">First sample name</em></dt>
<dd>Name for sample 1 to appear in plots legends</dd>
<dt><em class="guilabel">Second sample (counts)</em></dt>
<dd>Optional. If a second sample is available, this file should contain the same information as in <em class="guilabel">First sample</em>, but for the second sample. Mark the <em class="guilabel">Compare with other sample</em> box to enable this option.</dd>
<dt><em class="guilabel">Second sample name</em></dt>
<dd>Name for second sample to appear in plots legends.</dd>
<dt><em class="guilabel">Count threshold</em></dt>
<dd>A feature is considered as detected if the corresponding number of counts isgreater than this count threshold. By default, the theshold value is set to 5 counts.</dd>
<dt><em class="guilabel">Info File</em></dt>
<dd>Optional. File containing the biological classification of features in the count files. The info file must be a tab-delimited txt file, with the features names or IDs in the first column and the biological group (e.g. the biotype field from Biomart in the Ensembl database) in the second column. Again, the file must not contain any header or column names. If this file is provided, further exploratory plots can be generated to evaluate characteristics of the sample such us the kind of features being detected or the counts distribution for detected features. Please, make sure that the features IDs are the same in the <em class="guilabel">Info file</em> and in the <em class="guilabel">Count file</em>.</dd>
<dt><em class="guilabel">Species</em></dt>
<dd>Optional. If the Info File is not given by the user, Qualimap provides the Ensembl biotype classification for certain species (human and  mouse in Qualimap version 1.0), whenever the features names in the counts file are the Ensembl gene or transcripts IDs (e.g. ENSG00000251282 or ENST00000508921). If so, mark the box to enable this option and select the species.</dd>
</dl>
</div>
<div class="section" id="id5">
<h3>Output<a class="headerlink" href="#id5" title="Permalink to this headline">¶</a></h3>
<p><em class="guilabel">Global Saturation</em></p>
<blockquote>
<div><p>This plot provides information about the level of saturation in the sample, so it helps the user to decide if more sequencing is needed or if no many more features will detected when increasing the number of reads. These are some tips for the interpretation of the plot:</p>
<ul class="simple">
<li>The increasing sequencing depth of the sample is represented at the X-axis. The maximum value is the real sequencing depth of the sample(s). Smaller sequencing depths correspond to samples   randomly generated from the original sample(s).</li>
<li>The curve(s) is associated to the left Y-axis and represents the number of detected features when working with each of the sequencing depths in the X-axis. “Detected features” mean features with more than k counts, where k is the Count threshold chosen by the user.</li>
<li>The bars are associated to the right Y-axis and they represent the number of newly detected features when increasing the sequencing depth in one million reads at each sequencing depth value.</li>
</ul>
</div></blockquote>
<p>When an <em class="guilabel">Info File</em> is provided by the user or chosen from the ones supplied by Qualimap, a series of plots are additionally generated that are described next.</p>
<dl class="docutils">
<dt><em class="guilabel">Detection per group</em></dt>
<dd>This barplot allows the user to know which kind of features are being detected in their sample(s). The X-axis shows all the biological groups included in the Info file (or the biotypes supplied by Qualimap). The grey bar is the percentage of features in each biological group within the reference genome (or transcriptome, etc.). The striped color bar is the percentage detected in the sample with regard to the genome. The solid color bar is the percentage that the group (or biotype) represents in the total detected features in the sample.</dd>
<dt><em class="guilabel">Counts per group</em></dt>
<dd>A boxplot per each group (or biotype) describes the counts distribution for the detected features in that group.</dd>
<dt><em class="guilabel">Saturation per group</em></dt>
<dd>For each group (or biotype), a saturation plot is generated like the one described above.</dd>
<dt><em class="guilabel">Counts &amp; Sequencing Depth</em></dt>
<dd>For each group (or biotype), a plot is generated containing a boxplot with the distribution of counts at each sequencing depth. X-axis shows the increasing sequencing depths of randomly generated samples from the original one till the true sequencing depth is reached. This plot allows the user to see how the increase of sequencing depth is changing the expression level quantification.</dd>
</dl>
</div>
</div>
<div class="section" id="epigenomic-clustering">
<span id="epigenomic"></span><h2>Epigenomic (Clustering)<a class="headerlink" href="#epigenomic-clustering" title="Permalink to this headline">¶</a></h2>
<p>This analysis type allows to cluster the regions of interest in the input alignment based on the coverage. Typical usecase for such clustering can be evaluation of epigenomics experiment. For example due to epigenetics changes When analyzing promoter regions one should expect some genes to be activated and some silenced.</p>
<p>To perform this evaluation you need to provide alignment data ( both methylated and control) and list of transcript ids as BED annotation file.</p>
<div class="section" id="id6">
<h3>Input<a class="headerlink" href="#id6" title="Permalink to this headline">¶</a></h3>
<dl class="docutils">
<dt><em class="guilabel">Experiment ID</em></dt>
<dd>The experiment name</dd>
<dt><em class="guilabel">Alignment data</em></dt>
<dd><p class="first">Here you can provide your replicates to analyze. Each replicate includes sample file and a control file. For example, in an epigenomics experiment, the sample file could be the MeDIP-seq data and the control the non-encriched data (the so-called INPUT data). Thus, for each replicate the following information has to be provided:</p>
<dl class="docutils">
<dt><em class="guilabel">Replicate name</em></dt>
<dd>Name of the replicate</dd>
<dt><em class="guilabel">Sample file</em></dt>
<dd>Path to sample BAM file</dd>
<dt><em class="guilabel">Control file</em></dt>
<dd>Path to control BAM file</dd>
</dl>
<p class="last">To add a replicate click <em class="guilabel">Add</em> button. To remove a replicate select it and click <em class="guilabel">Remove</em> button. You can modify replicate by using <em class="guilabel">Edit</em> button.</p>
</dd>
<dt><em class="guilabel">Regions of interest</em></dt>
<dd><p class="first">Path to an annotation file in BED format which contains regions of interest.
The BED file should be a tab delimited text file with exactly 6 fields per line:</p>
<ul class="last simple">
<li>chromosome</li>
<li>start</li>
<li>end</li>
<li>name</li>
<li>score</li>
<li>strand (+ or -)</li>
</ul>
</dd>
<dt><em class="guilabel">Location</em></dt>
<dd>Relative location to analyze</dd>
<dt><em class="guilabel">Left offset</em></dt>
<dd>Offset in bp upstream the selected regions</dd>
<dt><em class="guilabel">Right offset</em></dt>
<dd>Offset in bp downstream the selected regions</dd>
<dt><em class="guilabel">Bin size</em></dt>
<dd>Can be thought as the resolution of the plot. Bins of the desired size will be computed and the information falling on each bin will be aggregated</dd>
<dt><em class="guilabel">Number of clusters</em></dt>
<dd>Number of groups that you the user wants to divide the data. Several values can be used by separting them with commas</dd>
<dt><em class="guilabel">Fragment length</em></dt>
<dd>Length of the fragments that were initially sequenced. All reads will be enlarged to this length.</dd>
<dt><em class="guilabel">Visualization type</em></dt>
<dd>You can visualize cluster using heatmaps or line-based graphs</dd>
</dl>
</div>
<div class="section" id="id7">
<h3>Output<a class="headerlink" href="#id7" title="Permalink to this headline">¶</a></h3>
<p>After the analysis is performed, the regions of interest are clustered in groups based on the coverage pattern. The output graph shows the coverage pattern for each cluster either as a heatmap or a line graph. There can be multiple graphs based on the number of clusters provided as input. The name of each graph consists of the experiment name and the number of clusters.</p>
<p>It is possible to export list of features beloning to the particular cluster. To do this use main menu item <em class="menuselection">File ‣ Export gene list</em> or context menu item <em class="menuselection">Export gene list</em>. After activating the item a dialog will appear where you can choose some specific cluster. One can either copy the list of features belonging to this cluster in the clipboard or export it to a text file.</p>
</div>
</div>
<div class="section" id="compute-counts">
<span id="id8"></span><h2>Compute counts<a class="headerlink" href="#compute-counts" title="Permalink to this headline">¶</a></h2>
<p>This tool allows to calculate how many reads belong to each region if interest in the alignment. To access the tool use menu item <em class="menuselection">Tools ‣ Compute counts</em>.</p>
<div class="section" id="id9">
<h3>Input<a class="headerlink" href="#id9" title="Permalink to this headline">¶</a></h3>
<dl class="docutils">
<dt><em class="guilabel">BAM file</em></dt>
<dd>Path to BAM alignment file</dd>
<dt><em class="guilabel">Annotation file</em></dt>
<dd>Path to GTF file containing regions of interest</dd>
<dt><em class="guilabel">Protocol</em></dt>
<dd><p class="first">Three options are avalalbe:</p>
<dl class="last docutils">
<dt><em class="dfn">non-strand-specific</em></dt>
<dd>Feature is counted independent of strand</dd>
<dt><em class="dfn">forward-stranded</em></dt>
<dd>Feature is counted only if it has the same strand as the read</dd>
<dt><em class="dfn">reverse-strand</em></dt>
<dd>Feature is counted only if the it has the strand reverse to the one of the read</dd>
</dl>
</dd>
<dt><em class="guilabel">Feature type</em></dt>
<dd>Third column of the GTF file. Only features of this particular type are counted.</dd>
<dt><em class="guilabel">Feature name</em></dt>
<dd>The name of the feature to be counted.</dd>
<dt><em class="guilabel">Output</em></dt>
<dd>Path to the file which will contain output.</dd>
<dt><em class="guilabel">Save computation summary</em></dt>
<dd>This option controls whether to save overall computation statistics.</dd>
</dl>
</div>
</div>
</div>


          </div>
        </div>
      </div>
        </div>
        <div class="sidebar">
          <h3>Table Of Contents</h3>
          <ul class="current">
<li class="toctree-l1"><a class="reference internal" href="intro.html">Introduction</a></li>
<li class="toctree-l1"><a class="reference internal" href="workflow.html">Workflow</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="">Analysis types</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#genomic">Genomic</a></li>
<li class="toctree-l2"><a class="reference internal" href="#rna-seq">RNA-seq</a></li>
<li class="toctree-l2"><a class="reference internal" href="#epigenomic-clustering">Epigenomic (Clustering)</a></li>
<li class="toctree-l2"><a class="reference internal" href="#compute-counts">Compute counts</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="command_line.html">Command line interface</a></li>
</ul>

          <h3 style="margin-top: 1.5em;">Search</h3>
          <form class="search" action="search.html" method="get">
            <input type="text" name="q" size="18" />
            <input type="submit" value="Go" />
            <input type="hidden" name="check_keywords" value="yes" />
            <input type="hidden" name="area" value="default" />
          </form>
          <p class="searchtip" style="font-size: 90%">
            Enter search terms or a module, class or function name.
          </p>
        </div>
        <div class="clearer"></div>
      </div>
    </div>

    <div class="footer-wrapper">
      <div class="footer">
        <div class="left">
          <a href="workflow.html" title="Workflow"
             >previous</a> |
          <a href="command_line.html" title="Command line interface"
             >next</a>
        </div>

        <div class="right">
          
    <div class="footer">
    </div>
        </div>
        <div class="clearer"></div>
      </div>
    </div>

  </body>
</html>